Add support for starting the system from USB media

150balbes opened the pull request.

150balbes committed:
Merge pull request #1 from hardkernel/odroidn2-v2015.01

20190531

150balbes committed:
add support start usb

150balbes committed:
add support start usb

Change-Id: I790a34ed734bdbc8206c73e986f793c37fe6722b

tobetter commented on include/configs/odroid-g12-common.h:
```diff
@@ -94,15 +94,16 @@
 #endif
 
-#define ENV_MMC_LIST_DEFAULT			"mmc_list=0 1\0"
+#define ENV_MMC_LIST_DEFAULT			"mmc_list=1 0\0"
```
We've been requested to swap the boot order from SD to eMMC, but we'd like to keep them eMMC to SD by default. Instead, we are happy to provide simple instruction to swap them by a user in Linux userspace or Android using a tool and preparing to release them by the end of next week.

150balbes replied:
To set the launch order, you can use a simple text file that will quickly change the launch order as desired by a specific user.

https://github.com/150balbes/Build-Armbian/blob/master/packages/bsp/odroid-n2/set_boot_order.ini

tobetter replied:
> To set the launch order, you can use a simple text file that will quickly change the launch order as desired by a specific user.
> 
> https://github.com/150balbes/Build-Armbian/blob/master/packages/bsp/odroid-n2/set_boot_order.ini

I know, that's what I intended to add the value "mmc_list" but not improved yet. I've guided the similar instruction to many users through the forum or PM but some of them still having a difficulty to follow it.

150balbes replied:
To use it, rename it (boot.ini) and write to any media (USB or SD card). By the way, I would recommend adding a mandatory shutdown of the system after the update procedure to the Piteboot update procedure. Now this procedure is repeated cyclically and you need to catch the moment and turn off N2.

150balbes commented:
For proper operation of the startup with USB in files (boot.ini) and (boot.scr) you need to add this information. The new system is compatible with the old files that are used in the old images (the system will start normally with SD\eMMC).

https://github.com/150balbes/Build-Armbian/blob/master/config/bootscripts/boot-odroid-n2.ini#L98-L103

https://github.com/150balbes/Build-Armbian/blob/master/config/bootscripts/boot-odroid-n2-mainline.ini#L12-L17

tobetter commented on include/configs/odroid-g12-common.h:
```diff
@@ -126,6 +127,40 @@
+                "status 'okay'; fi; bootm ${preloadaddr} ${initrd_high} ${dtb_mem_addr};" \
+                "\0"\
+        "boot_order=" \
+            "boot_usb boot_usb_distro boot_mmc boot_mmc_distro boot_rawimage boot_pxe boot_spi" \
```
As described above regarding boot order of MMC cards, we also like to keep the current boot order, but willing to offer the capability to change it by the user.

150balbes replied:
You are free to change this PR at your discretion. For me, it is important to be able to easily activate any desired user order in the SPI (by simply running a text file that is easy to edit to fit your needs).

tobetter replied:
Thanks again, actually I am doing a similar thing to offer a method to change the boot order in U-boot but not have spent much time for some other work. And I am planning to keep doing so in Petitboot by mid of June, but I am feeling that I would provide some way what you did soon. :)

tobetter reviewed (commented):
Thank you for your PR, we are thinking of the boot order. But since we are getting different requests from different users, we prefer to keep the current booting order but as I described in the review, we are willing to provide the instruction to customize the boot order.

tobetter commented:
> For proper operation of the startup with USB in files (boot.ini) and (boot.scr) you need to add this information. The new system is compatible with the old files that are used in the old images (the system will start normally with SD\eMMC).
> 
> https://github.com/150balbes/Build-Armbian/blob/master/config/bootscripts/boot-odroid-n2.ini#L98-L103
> 
> https://github.com/150balbes/Build-Armbian/blob/master/config/bootscripts/boot-odroid-n2-mainline.ini#L12-L17

Looks good to me, one suggestion in your change is to replace 'fatload' to 'load'. The command 'load' can manage the partition types of 'FAT32/64' and 'EXT2/3/4'.

150balbes commented:
That's reasonable, I'll change it in my files. By the way, I wouldn't spend a lot of time on Piteboot, u-boot-2019 for N2 with HDMI output support and direct systems intake control for startup should be out soon. :)

tobetter commented:
> That's reasonable, I'll change it in my files. By the way, I wouldn't spend a lot of time on Piteboot, u-boot-2019 for N2 with HDMI output support and direct systems intake control for startup should be out soon. :)

Wow, that's awesome. Glad to hear the mainline U-boot can run on ODROID-N2 sometime soon. :)

tobetter committed:
ODROID-C3/N2: revert the boot order of MMC (eMMC->SD)

Change-Id: I7fe13d45a655bd94a1a7351262b2e72d912ffb7e
Signed-off-by: Dongjin Kim <tobetter@gmail.com>

tobetter committed:
ODROID-C3/N2: refactoring the boot devices, scripts and order

This patch is to refactor the boot method defining the boot command per
device type, mmc/usb/spi/misc, and boot scripts, boot.ini/boot.scr. Also,
they can be customized its order by device types and device number.

Boot order per devices can be defined by 'boot_order';
	boot_order="mmc rawimage usb pxe spi"

In case multiple devices can be existed, they can be order by the device
numbers. For example;
	mmc_list="0 1"
	usb_list="0 1 2 3"

Change-Id: I3a0fad4915a4beb2039d1a1f087e99c91e3e5a31
Signed-off-by: Dongjin Kim <tobetter@gmail.com>

tobetter committed:
ODROID-C3/N2: add to able to write to ext4 file system

Change-Id: Id606c32388edea649ba866642ac4e84e3477b17b
Signed-off-by: Dongjin Kim <tobetter@gmail.com>

tobetter committed:
ODROID-C3/N2: config: update the boot command for SPI

Change-Id: Ief95d7c6e9d137be75bb6109346ac49695b31b25
Signed-off-by: Dongjin Kim <tobetter@gmail.com>

150balbes committed:
v20190604

150balbes committed:
Merge pull request #2 from tobetter/odroidn2-v2015.01

Odroidn2 v2015.01

150balbes committed:
Merge pull request #3 from hardkernel/odroidn2-v2015.01

20190605

150balbes committed:
fix

150balbes commented:
Corrected version.

150balbes commented:
Hi  @tobetter . 
Are you planning to release a version of u-boot SPI (image) with USB support ?

tobetter commented:
@150balbes No, I am not considering it at all.
